server.go

```diff
@@ -0,0 +1,264 @@
+	s.Listener = listener
+	s.serving = true
+	size := s.args.MessageMaxSize
+	if size < recvBufferSize {
+		size = recvBufferSize
+	}
+
+	for {
+		conn, err := s.transport.Listen()
+		if !s.serving {
+			return nil
+		}
+		if err != nil {
+			if e, ok := err.(net.Error); ok && e.Temporary() {
+				continue
+			}
+			return err
+		}
+
+		go func(conn interface{}) {
+			defer s.transport.Close(conn)
+			buf := make([]byte, size)
+			for {
+				_, src, msg, err := s.transport.Read(conn, buf)
+				if _, ok := err.(net.Error); ok {
+					if s.serving {
+						s.logf("trap: failed to read packet: %v", err)
+					}
+					return
+				}
+
+				go s.handle(conn, msg, src, err)
+			}
+		}(conn)
+	}
+}
+
+func (s *TrapServer) Close() error {
+	s.serving = false
+	return s.transport.Close(nil)
+}
+
+// handle a newly received trap
+func (s *TrapServer) handle(conn interface{}, msg message, src net.Addr, err error) {
+	defer func() {
+		if err := recover(); err != nil {
+			const size = 64 << 10
+			logBuf := make([]byte, size)
+			logBuf = logBuf[:runtime.Stack(logBuf, false)]
+			s.logf("trap: panic while receiving %v: %v\n%s", src, err, logBuf)
+
+		}
+	}()
+
+	l := s.Listener
+	if l == nil {
+		s.logf("trap: listener is not attached and trap information cannot be dispatched.")
+		return
+	}
+
+	var pdu Pdu
+	var sec security
+	if msg != nil {
+		if v := msg.Version(); v == V2c {
+			if sec = s.secs.Lookup(msg); sec != nil {
+				pdu, err = s.mp.PrepareDataElements(sec, msg, nil)
+			} else {
+				err = &MessageError{
+					Message: "Authentication failure",
+					Detail:  fmt.Sprintf("Message - [%s]", msg),
+				}
+			}
+		} else {
+			err = &MessageError{
+				Message: fmt.Sprintf("Unsupported SNMP version: %s", v),
+				Detail:  fmt.Sprintf("Message - [%s]", msg),
+			}
+		}
+	}
+
+	if pdu != nil {
+		switch t := pdu.PduType(); t {
+		case SNMPTrapV2, InformRequest:
+		default:
+			err = &MessageError{
+				Message: fmt.Sprintf("Invalid PduType: %s ", t),
+				Detail:  fmt.Sprintf("Message - [%s]", msg),
+			}
+			pdu = nil
+		}
+	}
+
+	l.OnTRAP(&TrapRequest{Pdu: pdu, Source: src, Error: err})
+
+	if pdu != nil && pdu.PduType() == InformRequest {
+		if err = s.informResponse(conn, src, sec, msg); err != nil && s.serving {
+			s.logf("trap: failed to send response %v: %v", src, err)
+		}
+	}
+}
+
+func (s *TrapServer) informResponse(conn interface{}, src net.Addr, sec security, msg message) error {
+	respPdu := NewPduWithVarBinds(msg.Version(), GetResponse, msg.Pdu().VarBinds())
+	respMsg, err := s.mp.PrepareResponseMessage(sec, respPdu, msg)
+	if err != nil {
+		return err
+	}
+	pkt, err := respMsg.Marshal()
+	if err != nil {
+		return err
+	}
+	return s.transport.Write(conn, pkt, src)
+}
+
+func (s *TrapServer) logf(format string, args ...interface{}) {
+	if l := s.ErrorLog; l != nil {
+		l.Printf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
+
+func NewTrapServer(args ServerArguments) (*TrapServer, error) {
```

Comment:
As It's part of the public API, it could be documented. 

```go
// NewTrapServer starts and returns a new Server and is using server arguments for configuration.
// The caller should call Close when finished, to shut it down.
func NewTrapServer(args ServerArguments) (*TrapServer, error) {
  ... 
}
```